### ISSUE TITLE
Bug: <Shard>.eval(() => someFunction) returns infinite promise

### ISSUE THREAD

**Issue by Mirdukkk** [discordjs/discord.js#6626]
### Issue description
When the function is used in [\<Shard\>](https://github.com/discordjs/discord.js/blob/main/src/sharding/Shard.js).eval(), it returns infinite promise.

Steps to reproduce:
1. Create sharding manager
2. Create some shards
3. Get any shard and try to .eval(() => something) using function

You will get approximately this result:
![image](https://user-images.githubusercontent.com/44965055/132961430-075e3fcc-30c6-4448-85ef-fe9660f193b5.png)

But if you pass a string instead of a function to eval, it will work:
![image](https://user-images.githubusercontent.com/44965055/132961496-174581db-902a-45fe-9c86-e9a094881777.png)

Notes: 
* The 'mode' of the sharding manager does not affect whether this bug exists or not.
* \<ShardingManager\>.broadcastEval(() => someFunction, { shard: 0 }) works fine.
* \<Shard\>.eval('string') works fine.
* The 'Discordoo' package used in the screenshots does not affect whether this bug exists or not.
* The number of shards does not affect whether this bug exists or not.

### Codesample

```typescript
// manager.js
const Discord = require('discord.js')

const manager = new Discord.ShardingManager('./shard.js', {
  totalShards: 1,
  token: process.env.TOKEN
})

manager.spawn()
  .then(shards => {
    console.log('getting guilds...')
    shards.get(0).eval(client => client.guilds.cache.size)
      .then(guilds => console.log('guilds:', guilds))
  })

// shard.js
const Discord = require('discord.js')

const client = new Discord.Client({
  intents: [ 'GUILDS' ]
})

client.login(process.env.TOKEN)
  .then(() => {
    console.log('ready!')
  })
```

### discord.js version

13.1.0

### Node.js version

16.6.0

### Operating system

Windows 10 Dev 21390.2025

### Priority this issue should have

Low (slightly annoying)

### Which gateway intents are you subscribing to?

GUILDS

### I have tested this issue on a development release

discord.js@13.2.0-dev.1631361856.be8912a

**Comment by DTrombett** [comment]
The eval with string shouldn't work anymore. Log your version using `npm ls discord.js`.

**Comment by ImRodry** [comment]
> The eval with string shouldn't work anymore. Log your version using `npm ls discord.js`.

Yes it does
![image](https://user-images.githubusercontent.com/38259440/132962308-8f8362d2-1d37-4762-a011-47f8876603cf.png)

**Comment by Mirdukkk** [comment]
> The eval with string shouldn't work anymore. Log your version using npm ls discord.js.

Here is the version from the log.
![image](https://user-images.githubusercontent.com/44965055/132962320-a6a3b3d1-d872-4bef-95c4-50ae796b5c01.png)

**Comment by comic-dev** [comment]
I've looked trough the `<Shard>.eval()` method and it seems to never resolve after sending the code to the shard. I've never helped with actual discord.js source code bugs so I don't know if this is the actual cause of the problem.

**Comment by DTrombett** [comment]
It's resolved here:
https://github.com/discordjs/discord.js/blob/88e2622ddebb8cacd264833ceb371e2fdbe512d5/src/sharding/Shard.js#L286

**Comment by DTrombett** [comment]
I'm not really sure that this is the cause but I think the problem is that `message._eval` will not be equal as `script` when it's a function since `_eval` is stringified before being sent:
https://github.com/discordjs/discord.js/blob/88e2622ddebb8cacd264833ceb371e2fdbe512d5/src/sharding/Shard.js#L291

If this is the problem the solution would be to move that `_eval` declaration up and check `message?._eval !== _eval` instead of `!== script`.

Summing up:
```diff
      const _eval = typeof script === 'function' ? `(${script})(this)` : script;
      const listener = message => {
        if (message?._eval !== _eval) return;
        child.removeListener('message', listener);
        this._evals.delete(script);
        if (!message._error) resolve(message._result);
        else reject(Util.makeError(message._error));
      };
      child.on('message', listener);

```

**Comment by DTrombett** [comment]
Tested and it works! I'll open a PR soon to fix this.
